packages/envd/internal/host/process.go

```diff
@@ -0,0 +1,130 @@
+package host
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/rs/zerolog"
+	"github.com/shirou/gopsutil/v4/process"
+)
+
+type ProcessState string
+
+const (
+	ProcessStateRunning ProcessState = "running"
+	ProcessStateExited  ProcessState = "exited"
+)
+
+type ProcessInfo struct {
+	State      ProcessState
+	PID        int32
+	Name       string
+	Cmdline    string
+	CreateTime int64
+}
+
+type ProcessEventHandler func(processInfo *ProcessInfo) error
+
+func getProcessInfo(pid int32) (*ProcessInfo, error) {
+	p, err := process.NewProcess(pid)
+	if err != nil {
+		return nil, err
+	}
+
+	name, err := p.Name()
+	if err != nil {
+		name = "unknown"
+	}
+
+	cmdline, err := p.Cmdline()
+	if err != nil {
+		cmdline = ""
+	}
+
+	createTime, err := p.CreateTime()
+	if err != nil {
+		createTime = 0
+	}
+
+	return &ProcessInfo{
+		PID:        pid,
+		Name:       name,
+		Cmdline:    cmdline,
+		CreateTime: createTime,
+	}, nil
+}
+
+func MonitorProcesses(ctx context.Context, logger *zerolog.Logger, interval time.Duration, processEventHandlers ...ProcessEventHandler) {
+	knownProcesses := make(map[string]*ProcessInfo)
+
+	// Get initial process list
+	initialPids, err := process.Pids()
+	if err != nil {
+		logger.Error().Err(err).Msg("Failed to get initial process list")
+	}
+
+	for _, pid := range initialPids {
+		if info, err := getProcessInfo(pid); err == nil {
+			key := deriveKey(pid, info.CreateTime)
+			knownProcesses[key] = info
+		} else {
+			logger.Error().Err(err).Msg("Failed to get process info")
+		}
+	}
+
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			currentPids, err := process.Pids()
+			if err != nil {
+				logger.Error().Err(err).Msg("Error getting current processes")
+				continue
+			}
```

Comment:
Just a quick note:

Can we watch for processes spawned by envd (run, command methods committed from our SDK)? I think monitoring all PIDs, except for some initial ones, will not work as systemd / other daemons can periodically spawn and kill processes, so current pids will be blown by some irrelevant processes.